### PR TITLE
Add the change_request_task module

### PR DESCRIPTION
**Pull request by MysteriousWolf**
##### SUMMARY
Adds change_request_task and change_request_task_info modules.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
- change_request_task and change_request_task_info

##### ADDITIONAL INFORMATION
The specification allows users to modify all [standard parameters](https://docs.servicenow.com/bundle/quebec-it-service-management/page/product/change-management/task/create-a-change-task.html) defined in the ServiceNow docs, as well as properties required to create and close tasks.

### CODE REVIEW

**Review by tadeboro — APPROVED** [review]

File: plugins/module_utils/validation.py [hunk 1]
```diff
@@ -38,3 +38,67 @@ def missing_from_params_and_remote(params, module_params, record=None):
             continue
         missing.append(p)
     return missing
+
+
+def check_value_compatibility(compatible_states, property_name, params, record=None):
```
Comment (tadeboro) [inline plugins/module_utils/validation.py]: As far as I can tell, this function is not used at all in the code. Is there a reason why we have it defined?

File: plugins/module_utils/validation.py [hunk 2]
```diff
@@ -38,3 +38,67 @@ def missing_from_params_and_remote(params, module_params, record=None):
...
+        else:
+            return False, property_value
+    # If the property is not in a compatible state
+    else:
+        return False, property_value
+
+
+def check_value_incompatibility(
+    incompatible_states, property_name, params, record=None
+):
```
Comment (tadeboro) [inline plugins/module_utils/validation.py]: Why does this function name hint at the state validation? As far as I can tell, all this function does is make sure certain parameter contains a valid value. And it does this in a quite convoluted way.

I would suggest renaming the function into something more sensible and simplifying it along those lines:

```
# Get value in effect
value = params[property_name]
if value is None:
    value = (record or {}).get(property_name)

# Make sure value is not in the set of invalid values
return value not in incompatible_states, value
```

File: plugins/modules/change_request_task.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+  - servicenow.itsm.sys_id
+  - servicenow.itsm.number
+
+seealso:
+  - module: servicenow.itsm.change_request_task_info
+
+options:
+  configuration_item:
+    description:
+      - The configuration item (CI) or service name that the change task applies to.
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py:38]: We should inform the user here then CI names are not guaranteed to be unique. If that is the case, the module will fail and do nothing.

File: plugins/modules/change_request_task.py [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+
+from ansible.module_utils.basic import AnsibleModule
+
+from ..module_utils import arguments, client, table, errors, utils, validation
+from ..module_utils.change_request_task import PAYLOAD_FIELDS_MAPPING
+
+DIRECT_PAYLOAD_FIELDS = (
+    "state",
+    "assigned_to",
+    "assignment_group",
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]: Is this really copy-paste parameter or should we do a lookup of the assignment group's ID?

File: plugins/modules/change_request_task.py [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+"""
+
+EXAMPLES = """
+- name: Create a change task
+  servicenow.itsm.change_request_task:
+    instance:
+      host: https://instance_id.service-now.com
+      username: user
+      password: pass
+
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]: Just drop this part in all other examples since 1) authentication is not that interesting once you saw it once and 2) this is not how instance data is usually passed to the modules.

File: plugins/modules/change_request_task.py [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+    if change:
+        table_client.delete_record("change_task", change, module.check_mode)
+        return True, None, dict(before=mapper.to_ansible(change), after=None)
+
+    return False, None, dict(before=None, after=None)
+
+
+def validate_params(params, change_request=None):
+    missing = []
+    if params.get("state") == "closed":
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]:
```suggestion
    if params["state"] == "closed":
```

File: plugins/modules/change_request_task.py [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+        missing.extend(
+            validation.missing_from_params_and_remote(
+                ("close_code", "close_notes"), params, change_request
+            )
+        )
+
+    if params.get("on_hold"):
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]:
```suggestion
    if params["on_hold"]:
```

File: plugins/modules/change_request_task.py [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+        compatibility = validation.check_value_incompatibility(
+            ("pending", "canceled", "closed"), "state", params, change_request
+        )
+        if compatibility[0] is False:
+            raise errors.ServiceNowError(
+                'Cannot put a task in state "{}" on hold'.format(compatibility[1])
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]: This error message is bad because it will mislead users in certain scenarios. For example, if the user works on an existing task that is already on hold and tries to change its state to `canceled`, the error message makes no sense since our poor user is not trying to put anything on hold at all.

**Review by tadeboro — CHANGES_REQUESTED** [review]
I did my initial review and I found a few areas that could use some improvement, especially around usability (providing misleading error messages is a no-go).

File: plugins/module_utils/validation.py (hunk 1 above)
Comment (MysteriousWolf) [reply plugins/module_utils/validation.py]: I put it there mostly for future use since it is very similar to the other incompatibility check, but I agree that it could be omitted since it possibly won't be used anywhere anyway.

File: plugins/modules/change_request_task.py (hunk 8 above)
Comment (MysteriousWolf) [reply plugins/modules/change_request_task.py]: Not only that, but it doesn't even check whether the task may already be on hold and the user wants to set the state to something incompatible. Working on a fix that should cover both cases and provide better feedback.

File: plugins/modules/change_request_task.py [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+  - Miha Dolinar (@mdolin)
+  - Tadej Borovsak (@tadeboro)
+
+short_description: Manage ServiceNow change request tasks
+
+description:
+  - Create, delete or update a ServiceNow change request tasks.
+  - For more information, refer to the ServiceNow change management documentation at
+    U(https://docs.servicenow.com/bundle/paris-it-service-management/page/product/change-management/concept/c_ITILChangeManagement.html).
+version_added: 1.0.0
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]:
```suggestion
version_added: 1.2.0
```

File: plugins/modules/change_request_task.py [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+  - servicenow.itsm.number
+
+seealso:
+  - module: servicenow.itsm.change_request_task_info
+
+options:
+  configuration_item:
+    description:
+      - The configuration item (CI) or service name that the change task applies to.
+      - Note that contrary to configuration_item_id, configuration item names may not uniquely identify a record.
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]:
```suggestion
      - Note that contrary to I(configuration_item_id), configuration item names may not uniquely identify a record.
```

File: plugins/modules/change_request_task.py [hunk 11]
```diff
@@ -0,0 +1,434 @@
...
+    type: str
+  configuration_item_id:
+    description:
+      - The configuration item (CI) or service ID that the change task applies to.
+      - Mutually exclusive with I(configuration_item).
+    type: str
+  change_request_number:
+    description:
+      - I(number) of the change request this task belongs to.
+      - Note that contrary to change_request_id, change request number may not uniquely identify a record.
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]:
```suggestion
      - Note that contrary to I(change_request_id), change request number may not uniquely identify a record.
```

File: plugins/modules/change_request_task.py [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+                'Cannot put a task in state "{}" on hold'.format(state)
+            )
+    elif (change_request or {}).get("on_hold"):
+        compatible, state = validation.check_value_incompatibility(
+            ("pending",), "state", params, change_request
+        )
+        if not compatible:
+            raise errors.ServiceNowError(
+                'Cannot change task state to "{}" when on hold'.format(state)
+            )
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]: Why is this different from the previous branch? I would assume the same validation criteria applies here.

File: plugins/modules/change_request_task.py [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+def is_superset_with_date(superset, candidate):
+    for k, v in candidate.items():
+        if k not in superset or superset[k] != v:
+            if not (
+                (k == "planned_start_date" or k == "planned_end_date")
+                and (str(superset[k]) != "" and str(v) != "")
+                and str(superset[k]).replace("T", " ") == str(v).replace("T", " ")
+            ):
+                return False
+    return True
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]: This feels a bit awkward. Since we are only interested in adjusting the result in case any of the two date fields are different in representation but actually represent the same point in time, we should probably do something like this:

```suggestion
def is_superset_with_date(superset, candidate):
    datetime_fields = ("planned_start_date", "planned_end_date")
    
    for k in datetime_fields:
        if k in candidate and superset.get(k, "").replace("T", " ") != candidate[k].replace("T", " "):
            return False
    
    return utils.is_superset(superset, dict((k, v) for k, v in candidate.items() if k not in datetime_fields))
```

**Review by tadeboro — CHANGES_REQUESTED** [review]

File: plugins/modules/change_request_task.py (hunk 12 above)
Comment (MysteriousWolf) [reply plugins/modules/change_request_task.py]: When testing the plugin, I found out that ServiceNow allows changing task states to closed or cancelled even if it's currently on hold. Contrary to that, it does not allow you to set it on hold when in those two states or when it's still "pending", which is also reflected in the snow UI, so I separated them into two different branches.

Comment (tadeboro) [reply plugins/modules/change_request_task.py]: > When testing the plugin, I found out that ServiceNow allows changing task states to closed or cancelled even if it's currently on hold. Contrary to that, it does not allow you to set it on hold when in those two states or when it's still "pending", which is also reflected in the snow UI, so I separated them into two different branches.

But if I understand this correctly, this is not what this check enforces. Let us assume that we have an existing task that has `state=in_progress` and `on_hold=true`. If I run the following task, the validation will abort the task execution:

```
- servicenow.itsm.change_request_task:
    number: 1234
    state: closed
    on_hold: true
```

But if I understand you correctly, this task should succeed because transitioning from `in_progress` to `closed`is valid even when the task is on hold.

File: plugins/module_utils/validation.py [hunk 14]
```diff
@@ -38,3 +38,19 @@ def missing_from_params_and_remote(params, module_params, record=None):
...
+    incompatible_values, property_name, params, record=None
+):
+    """
+    Given a list of incompatible values, a property name, params dict and a ServiceNow record, returns the check
+    result and value of the checked property. Check result is True if property values aren't in the incompatible
+    value list and False otherwise. The function assumes that new parameters override the record.
+    """
+
+    # Get value in effect
+    value = params[property_name] or (record or {}).get(property_name)
```
Comment (tadeboro) [inline plugins/module_utils/validation.py]: Is this really what we want here? `None` is not the only falsy thing in Python, so this has the potential to give us trouble. For example, if the user explicitly sets one of the parameters to `""`, this line will "ignore" the empty string and instead return whatever is in the record.

And while on its own this behavior would note problematic, combined with https://github.com/xlab-steampunk/servicenow.itsm/blob/48b521fd7e83a22994def9b010b674150c99e8a8/plugins/module_utils/utils.py#L11-L19 does make things bad because the validation and payload building now behave differently.

File: tests/unit/plugins/module_utils/test_validation.py [hunk 15]
```diff
@@ -73,3 +73,109 @@ def test_invalid_not_a_subset(self):
...
+            ),  # state not set, record doesn't exist
+        ],
+    )
+    def test_state_compatible(
+        self, incompatible_states, property_name, params, record, resulting_property
+    ):
+        result = validation.check_value_incompatibility(
+            incompatible_states, property_name, params, record
+        )
+        print(result)
```
Comment (tadeboro) [inline tests/unit/plugins/module_utils/test_validation.py]:
```suggestion
```

File: tests/unit/plugins/module_utils/test_validation.py [hunk 16]
```diff
@@ -73,3 +73,109 @@ def test_invalid_not_a_subset(self):
...
+            ),  # state used to be incompatible, is now set correctly
+            (("a",), "p", dict(p=None), dict(), None),  # state not set
+            (
+                ("a",),
+                "p",
+                dict(p=None),
+                None,
+                None,
+            ),  # state not set, record doesn't exist
+        ],
```
Comment (tadeboro) [inline tests/unit/plugins/module_utils/test_validation.py]: We should also add some test samples for edge cases, for example, where the `params["p"] = ""`. Also, testing things on other types of values apart from strings should be done (we do not need new tests for this, we can use ints in some other tests, again making sure things like `0` that are falsy produce the correct result).

File: tests/unit/plugins/module_utils/test_validation.py [hunk 17]
```diff
@@ -73,3 +73,109 @@ def test_invalid_not_a_subset(self):
...
+                "a",
+            ),  # state already incorrect and not corrected
+            (
+                ("a",),
+                "p",
+                dict(p="a"),
+                dict(p="b"),
+                "a",
+            ),  # state correct in the record but not in the module
+        ],
```
Comment (tadeboro) [inline tests/unit/plugins/module_utils/test_validation.py]: The same comments apply here as for the previous test.

File: plugins/modules/change_request_task.py [hunk 18]
```diff
@@ -0,0 +1,432 @@
...
+      - This field has to be set either in the record or here.
+    type: str
+  description:
+    description:
+      - A detailed description of the task.
+      - This field has to be set either in the record or here.
+    type: str
+  on_hold:
+    description:
+      - A change task can be put on hold when I(state) is not in the C(pending), C(canceled), or C(closed)
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]: Is this correct? If I remember correctly, we talked about this and concluded that we could not put a task on hold when it is pending, canceled, or closed.

File: plugins/modules/change_request_task.py [hunk 19]
```diff
@@ -0,0 +1,432 @@
...
+      - Required if change task's I(on_hold) value will be C(true).
+    type: str
+  planned_start_date:
+    description:
+      - The date you plan to begin working on the task.
+    type: str
+  planned_end_date:
+    description:
+      - The date the change task is planned to be completed.
+      - If the task I(type) is C(implementation), the I(planned_start_date) and I(planned_end_date) values
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]: The comment about the `planned_start_date` parameter should live in the previous description.

File: plugins/modules/change_request_task.py [hunk 20]
```diff
@@ -0,0 +1,432 @@
...
+    change = table_client.get_record("change_task", query)
+
+    if change:
+        table_client.delete_record("change_task", change, module.check_mode)
+        return True, None, dict(before=mapper.to_ansible(change), after=None)
+
+    return False, None, dict(before=None, after=None)
+
+
+def validate_params(params, change_request=None):
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]:
```suggestion
def validate_params(params, change_task=None):
```

File: tests/unit/plugins/modules/test_change_request_task.py [hunk 21]
```diff
@@ -0,0 +1,520 @@
...
+        table_client.create_record.return_value = dict(
+            change_task_type="planning",
+            state="1",
+            short_description="sd",
+            description="description",
+            number="CTASK0000001",
+        )
+
+        result = change_request_task.ensure_present(module, table_client)
+        print(result)
```
Comment (tadeboro) [inline tests/unit/plugins/modules/test_change_request_task.py]:
```suggestion
```

File: tests/unit/plugins/modules/test_change_request_task.py [hunk 22]
```diff
@@ -0,0 +1,520 @@
...
+        table_client.get_record.side_effect = [
+            {"sys_id": "c248952584a34ae1a851a38d7fc08fcf"},
+            {"sys_id": "e361760abb09450da835b5e4f2271dcf"},
+            {"sys_id": "4488052c5f5248f8b787ec9df5459c09"},
+            {"sys_id": "18fee81900824bfeb78e8c9062fb8f06"},
+        ]
+
+        result = change_request_task.build_payload(module, table_client)
+
+        print(result)
```
Comment (tadeboro) [inline tests/unit/plugins/modules/test_change_request_task.py]:
```suggestion
```

File: tests/unit/plugins/modules/test_change_request_task.py [hunk 23]
```diff
@@ -0,0 +1,520 @@
...
+            ),
+        )
+        table_client.get_record.side_effect = [
+            {"sys_id": "c248952584a34ae1a851a38d7fc08fcf"},
+            {"sys_id": "e361760abb09450da835b5e4f2271dcf"},
+        ]
+
+        result = change_request_task.build_payload(module, table_client)
+
+        print(result)
```
Comment (tadeboro) [inline tests/unit/plugins/modules/test_change_request_task.py]:
```suggestion
```

File: tests/unit/plugins/modules/test_change_request_task.py [hunk 24]
```diff
@@ -0,0 +1,520 @@
...
+    )
+    def test_validate_params_put_on_hold(self, state, valid):
+        params = self.VALID_PARAMS_HOLD.copy()
+        params["state"] = state
+
+        if valid:
+            change_request_task.validate_params(self.VALID_PARAMS_HOLD)
+        else:
+            with pytest.raises(errors.ServiceNowError, match=state):
+                change_request_task.validate_params(params)
```
Comment (tadeboro) [inline tests/unit/plugins/modules/test_change_request_task.py]: Just split those two categories of problems into two test functions. Having conditional logic in unit tests is not the best design. In an ideal world, each unit test is composed of three sections with as few logical constructs as possible: environment setup (if needed), function call, and assertions (if needed).

File: tests/unit/plugins/modules/test_change_request_task.py [hunk 25]
```diff
@@ -0,0 +1,520 @@
...
+        params = self.VALID_PARAMS.copy()
+        params["state"] = state
+
+        if valid:
+            change_request_task.validate_params(
+                self.VALID_PARAMS_HOLD, self.VALID_PARAMS_HOLD
+            )
+        else:
+            with pytest.raises(errors.ServiceNowError, match=state):
+                change_request_task.validate_params(params, self.VALID_PARAMS_HOLD)
```
Comment (tadeboro) [inline tests/unit/plugins/modules/test_change_request_task.py]: See the previous comment.

File: tests/unit/plugins/modules/test_change_request_task.py [hunk 26]
```diff
@@ -0,0 +1,520 @@
...
+                dict(planned_end_date="2021-07-09 08:40:33"),
+                dict(planned_end_date="2021-07-09T08:40:33"),
+            ),
+            (
+                dict(planned_end_date="2021-07-09 08:40:33"),
+                dict(planned_end_date="2021-07-09 08:40:33"),
+            ),
+        ],
+    )
+    def test_valid_dates(self, record, params):
```
Comment (tadeboro) [inline tests/unit/plugins/modules/test_change_request_task.py]:
```suggestion
    def test_same_point_in_time(self, record, params):
```

File: tests/unit/plugins/modules/test_change_request_task.py [hunk 27]
```diff
@@ -0,0 +1,520 @@
...
+        assert change_request_task.is_superset_with_date(record, params) is True
+
+    @pytest.mark.parametrize(
+        "record,params",
+        [
+            (
+                dict(planned_start_date="2021-07-09 08:40:33"),
+                dict(planned_start_date="2021-07-09 08:40:34"),
```
Comment (tadeboro) [inline tests/unit/plugins/modules/test_change_request_task.py:470]: This test case makes sure we detect if we are operating on different dates and should be in a dedicated test together with a companion end date test case.

File: tests/unit/plugins/modules/test_change_request_task.py [hunk 28]
```diff
@@ -0,0 +1,520 @@
...
+            ),
+            (
+                dict(planned_start_date=""),
+                dict(planned_start_date="2021-07-09 08:40:34"),
+            ),
+            (
+                dict(planned_start_date="2021-07-09 08:40:33"),
+                dict(planned_start_date=""),
+            ),
+            (dict(param="2021-07-09 08:40:33"), dict(param="2021-07-09T08:40:33")),
```
Comment (tadeboro) [inline tests/unit/plugins/modules/test_change_request_task.py]: We should move this test case to the `not_a_superset` test and its positive equivalent to the `valid_superset` test.

File: tests/unit/plugins/modules/test_change_request_task.py [hunk 29]
```diff
@@ -0,0 +1,520 @@
...
+                dict(planned_start_date="2021-07-09 08:40:33"),
+                dict(planned_start_date="2021-07-09 08:40:34"),
+            ),
+            (
+                dict(planned_start_date=""),
+                dict(planned_start_date="2021-07-09 08:40:34"),
+            ),
+            (
+                dict(planned_start_date="2021-07-09 08:40:33"),
+                dict(planned_start_date=""),
```
Comment (tadeboro) [inline tests/unit/plugins/modules/test_change_request_task.py:514]: We also need test cases with `None`s and absent parameters in the params dictionary.

File: plugins/modules/change_request_task_info.py [hunk 30]
```diff
@@ -0,0 +1,241 @@
...
+def run(module, table_client):
+    mapper = utils.PayloadMapper(PAYLOAD_FIELDS_MAPPING, module.warn)
+
+    return [
+        mapper.to_ansible(record)
+        for record in table_client.list_records(
+            "change_task",
+            {"sysparm_query": sysparms_query(module, table_client, mapper)}
+            if module.params["query"]
+            else utils.filter_dict(module.params, "sys_id", "number"),
```
Comment (tadeboro) [inline plugins/modules/change_request_task_info.py]: This multiline if expression should be a separate statement because it makes it hard to see the actual parameters.  And as a side benefit, we will also gain a small speed boost because we will only construct query once and not once per iteration.

**Review by tadeboro — CHANGES_REQUESTED** [review]
Overall, this PR is turning out to be a great contribution to the codebase. I did find some issues, but we are progressing really nicely here. Thank you, @MysteriousWolf, for your effort!

File: plugins/module_utils/validation.py (hunk 14 above)
Comment (MysteriousWolf) [reply plugins/module_utils/validation.py]: I agree. While it would work most of the time, it could cause issues when working with not None, falsy values.

File: plugins/modules/change_request_task.py (hunk 18 above)
Comment (MysteriousWolf) [reply plugins/modules/change_request_task.py]: It's worded a bit awkwardly. I'll put the negation sooner in the sentence to clear up the possible confusion.

File: tests/unit/plugins/modules/test_change_request_task.py (hunk 28 above)
Comment (MysteriousWolf) [reply tests/unit/plugins/modules/test_change_request_task.py]: I don't think the valid superset equivalent is necessary since we already test for valid dates in `test_same_point_in_time`, but I did move this specific case to `not_a_superset` test since it doesn't have anything to do with dates in the background.

File: plugins/modules/change_request_task.py [hunk 31]
```diff
@@ -0,0 +1,436 @@
...
+    "planned_end_date",
+    "close_code",
+    "close_notes",
+)
+
+
+def ensure_absent(module, table_client):
+    mapper = utils.PayloadMapper(PAYLOAD_FIELDS_MAPPING, module.warn)
+    query = utils.filter_dict(module.params, "sys_id", "number")
+    change = table_client.get_record("change_task", query)
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]:
```suggestion
    task = table_client.get_record("change_task", query)
```

Or `change_task`, whatever feels more natural. But `change` is not OK because it is misleading.

File: plugins/modules/change_request_task.py [hunk 32]
```diff
@@ -0,0 +1,436 @@
...
+                'Cannot put a task in state "{}" on hold'.format(state)
+            )
+    elif (change_task or {}).get("on_hold"):
+        compatible, state = validation.check_value_incompatibility(
+            ("pending",), "state", params, change_task
+        )
+        if not compatible:
+            raise errors.ServiceNowError(
+                'Cannot change task state to "{}" when on hold'.format(state)
+            )
```
Comment (tadeboro) [inline plugins/modules/change_request_task.py]: The logic behind this is not entirely clear to me. Let us assume that we have an existing task that has `state=in_progress` and `on_hold=true`. If I run the following task, the validation will abort the task execution:

```
- servicenow.itsm.change_request_task:
    number: 1234
    state: closed
    on_hold: true
```

But if I understand you correctly, this task should succeed because transitioning from `in_progress` to `closed` is valid even when the task is on hold.

**Review by tadeboro — CHANGES_REQUESTED** [review]
Almost there ;)

File: plugins/modules/change_request_task.py (hunk 32 above)
Comment (MysteriousWolf) [reply plugins/modules/change_request_task.py]: On one hand this is true, the task could be put on hold, but on the other, it doesn't make sense to change the state to closed and put the task on hold in the same Ansible task (snow UI doesn't allow you to do that either), but you can still close a task that is already on hold. If you closed it already, why put it on hold?

Comment (tadeboro) [reply plugins/modules/change_request_task.py]: The problem here is the misleading error message. When my example task fails, it will say that we cannot put the task on hold because of the incompatible state. This is not true since the task is already on hold and users will be angry at us because we are trying to do something that does not need to be done.

Ansible playbook is all about the final state and this makes a difference interpreting the actions required. My example does not say "move task to closed state and put it on hold." It says "make sure task is closed and on hold." And we should only complain if the final state is invalid or if the transition from the current to the end state is impossible. And neither is true in this example.

Comment (MysteriousWolf) [reply plugins/modules/change_request_task.py]: Do you think adding an exception for when the task isn't really put `on_hold`, but rather just making sure that it's already on hold is an option? In this case we could make sure it's on hold, but wouldn't really be consistent and could fail if the state needs changing.

Comment (tadeboro) [reply plugins/modules/change_request_task.py]: As far as I can tell, we need to handle the following cases:

 1. If the existing task is on hold and we are setting on hold state to false or not setting it at all, everything goes.
 2. If the existing task is on hold and we are also setting on hold to true, the final state cannot be pending (and the error should reflect that the state move is problematic, not the on hold status).
 3. If the existing task is not on hold and we are setting on hold state to false or not setting it at all, everything goes.
 4. If the existing task is not on hold and we are setting the on hold to true, the final state cannot be pending, canceled, closed.

Did I miss anything here?

Comment (MysteriousWolf) [reply plugins/modules/change_request_task.py]: At first I built this in a way to follow the UI constraints, but I did a bit more testing right now and the API doesn't seem to really care what combination of these parameters you give. I guess leaving out all the validation of these two parameters could also be an option.

Comment (tadeboro) [reply plugins/modules/change_request_task.py]: If this is the case, then yes, drop the validation. Simpler is better in most cases.

**Review by tadeboro — CHANGES_REQUESTED** [review]
Once we rebase this, we are ready to go.

**Review by tadeboro — APPROVED** [review]
